File: weasel.py
import string
import random

target = 'METHINKS IT IS LIKE A WEASEL'
rand_str = ''.join(random.choice(string.ascii_uppercase + ' ') for _ in range(len(target)))
mut = 5
copies = 100

def mod_rand_str(rand_str):
    new_str = rand_str
    for i in range(len(target)):
        probability = random.randint(0, 100)
        if probability <= mut:
            new_str = new_str[:i] + random.choice(string.ascii_uppercase + ' ') + new_str[i+1:]
    return new_str

def mod_rand_str_lock(rand_str):
    new_str = rand_str
    for i in range(len(target)):
        if new_str[i] != target[i]:
            probability = random.randint(0, 100)
            if probability <= mut:
                new_str = new_str[:i] + random.choice(string.ascii_uppercase + ' ') + new_str[i+1:]
    return new_str

def comp_str(rand_str):
    matches = 0
    for i in range(len(target)):
        if target[i] == rand_str[i]:
            matches += 1
    return matches

result_list = []
str_found = False
loop_number = 0
while not str_found:
    str_list = [''] * copies
    matches_list = [None] * copies
    for i in range(copies):
        str_list[i] = mod_rand_str(rand_str)
        matches_list[i] = comp_str(str_list[i])
    max_matches = max(matches_list)
    if max_matches == len(target):
        str_found = True
    rand_str = str_list[matches_list.index(max_matches)]
    result_list.append(str(loop_number) + ': ' + rand_str + ' | matches= ' + str(max_matches) +
                       ' | index in gen= ' + str(matches_list.index(max_matches)) )
    loop_number += 1

for i in result_list:
    print i

File: weasel.js
const target = 'METHINKS IT IS LIKE A WEASEL';
const mut = 5;
const copies = 100;
const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ ';

function genStr(){
  let str = '';
  for (let i = 0; i < target.length; i++){
    str += alphabet[Math.floor(Math.random() * alphabet.length)]
  }
  return str;
};

function compStr(str){
  let matches = 0;
  for (let i = 0; i < target.length; i++){
    if (str[i] === target[i]){
      matches += 1;
    }
  }
  return matches;
};

function modStr(str){
  let newStr = str;
  for (let i = 0; i < str.length; i++){
    const prob = Math.floor(Math.random() * 100);
    if (prob < mut){
      newStr = newStr.substring(0, i) + alphabet.charAt(Math.floor(Math.random() * alphabet.length))
              + newStr.substring(i+1, str.length);
    }
  }
  return newStr;
};

// the same func as modStr but letters which are correct can't mutate. To use this - change line 60
function modStrLock(str){
  let newStr = str;
  for (let i = 0; i < str.length; i++){
    if(newStr[i] !== target[i]){
      const prob = Math.floor(Math.random() * 100);
      if (prob < mut){
        newStr = newStr.substring(0, i) + alphabet.charAt(Math.floor(Math.random() * alphabet.length))
                + newStr.substring(i+1, str.length);
      }
    }
  }
  return newStr;
};

var resultList = [];
var startStr = genStr();
var strFound = false;
var loopNumber = 0;

while (!strFound){
  var strList = Array.from(Array(copies), x => '');
  var matchesList = Array.from(Array(copies), x => '');
  for (let i = 0; i < copies; i++){
    strList[i] = modStr(startStr);
    matchesList[i] = compStr(strList[i]);
  }
  maxMatch = Math.max.apply(null, matchesList);
  if (maxMatch === target.length){
    strFound = true;
  }
  startStr = strList[matchesList.indexOf(maxMatch)];
  resultList.push(loopNumber + ': ' + startStr + ' |matches: ' + maxMatch);
  loopNumber += 1;
};

for (let i = 0; i < resultList.length; i++){
  console.log(resultList[i]);
}
